sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala

```diff
@@ -293,4 +293,29 @@ private[spark] object SchemaUtils {
    * @return The escaped string.
    */
   def escapeMetaCharacters(str: String): String = SparkSchemaUtils.escapeMetaCharacters(str)
+
+  /**
+   * Checks if a given data type has a non-default collation string type.
+   */
+  def hasNonDefaultCollatedString(dt: DataType): Boolean = {
+    typeExistsRecursively(dt) {
+      case st: StringType => !st.isDefaultCollation
+      case _ => false
+    }
+  }
+
+  /**
+   * Recursively checks whether a given predicate holds true for any data type
+   * within the specified data type.
+   */
+  def typeExistsRecursively(dt: DataType)(f: DataType => Boolean): Boolean = dt match {
```

Comment:
isn't it just `DataType#existsRecursively`?

Reply:
🤦 yes, my bad sorry